File: netlogo-gui/src/main/prim/_bk.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.core.Let;
import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.MutableDouble;

// note that this and _fd are pretty much carbon copies of each other

public final strictfp class _bk
    extends Command
    implements org.nlogo.nvm.CustomAssembled {
  final Let let = new Let(null);

  @Override
  public org.nlogo.core.Syntax syntax() {
    return Syntax.commandSyntax
        (new int[]{Syntax.NumberType()},
            "-T--");
  }

  @Override
  public void perform(Context context) throws LogoException {
    perform_1(context, argEvalDoubleValue(context, 0));
  }

  public void perform_1(Context context, double d) {
    context.let(let, new MutableDouble(-d));
    context.ip = next;
  }

  public void assemble(org.nlogo.nvm.AssemblerAssistant a) {
    a.add(this);
    a.add(new _fdinternal(this));
  }
}

File: netlogo-headless/src/main/prim/_bk.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.core.{ Let, Token }
import org.nlogo.nvm.{ AssemblerAssistant, Command, Context, CustomAssembled, MutableDouble }

// note that this and _fd are pretty much carbon copies of each other

class _bk(_token: Token) extends Command with CustomAssembled {
  token_=(_token)

  // MethodRipper won't let us call a public method from perform_1() - ST 7/20/12
  private[this] val _let = Let(s"~${_token.text}_${_token.start}")

  def let = _let

  override def perform(context: Context): Unit = {
    perform_1(context, argEvalDoubleValue(context, 0))
  }

  def perform_1(context: Context, d: Double): Unit = {
    context.activation.binding.let(_let, new MutableDouble(-d))
    context.ip = next
  }

  def assemble(a: AssemblerAssistant): Unit = {
    a.add(this)
    a.add(new _fdinternal(this))
  }
}
